【Hackathon No.56 57 58 59】sparse elementwise add sub mul div

gsq7474741 opened the pull request:
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
OPs
### Describe
<!-- Describe what this PR does. -->
add paddle.incubate.sparse.add subtract multiply divide and c++ kernels
This PR is an entry point of Hackathon task No.56 57 58 59
Tasks: 
- https://github.com/PaddlePaddle/Paddle/issues/40275
- https://github.com/PaddlePaddle/Paddle/issues/40274
- https://github.com/PaddlePaddle/Paddle/issues/40273
- https://github.com/PaddlePaddle/Paddle/issues/40272

Related work:
- Docs: https://github.com/PaddlePaddle/docs/pull/4705
- RFCs: https://github.com/PaddlePaddle/community/pull/95
- RFCs: https://github.com/PaddlePaddle/community/pull/123
<!-- 
DO NOT merge this PR until part No.1 and No.2 are merged sequentially

Sub-PR: 
- PR 1：#42374 
- PR 2：#42375
- Docs: https://github.com/PaddlePaddle/docs/pull/4705-->

gsq7474741 committed:
add sparse elementwise

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

gsq7474741 committed:
add sparse elementwise v0.2

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

gsq7474741 committed:
add sparse elementwise v0.3

reset author

gsq7474741 committed:
add sparse elementwise v0.3

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	python/paddle/sparse/__init__.py

gsq7474741 committed:
add sparse elementwise v0.3

gsq7474741 committed:
add sparse elementwise v0.5

gsq7474741 committed:
add sparse elementwise v0.5

gsq7474741 committed:
add sparse elementwise v0.5

gsq7474741 committed:
add sparse elementwise v0.5

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

gsq7474741 committed:
add sparse elementwise v0.7

gsq7474741 committed:
add sparse elementwise v0.7

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,307 @@
+                                    const SparseCsrTensor& x,                  \
+                                    const SparseCsrTensor& y,                  \
+                                    SparseCsrTensor* out) {                    \
+    PADDLE_ENFORCE_EQ(x.dims(), y.dims());                                     \
```
需要给出校验的提示信息。

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,307 @@
+                                    SparseCsrTensor* out) {                    \
+    PADDLE_ENFORCE_EQ(x.dims(), y.dims());                                     \
+    const auto& n_row = x.dims()[0];                                           \
+    const auto& n_col = x.dims()[1];                                           \
```
当前的SparseCsrTensor支持2-D和3-D的矩阵，所以这里还需要考虑下3-D的输入。

zkh2016 commented on python/paddle/sparse/math.py:
```diff
@@ -0,0 +1,237 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
2020 -> 2022

zkh2016 commented on python/paddle/sparse/math.py:
```diff
@@ -0,0 +1,237 @@
+    # ), "Currently, sparse.add only support the input of SparseCsrTensor"
+    if x.is_sparse_coo() or y.is_sparse_coo():
+        if x.is_sparse_coo():
+            _x = x.to_sparse_csr()
```
1. 因为当前不支持csr的反向，所以不能够在python层进行转换，可以提供一个coo的api，在c++端进行转换。
2. 另外这里直接转成csr会有问题，SparseCsrTensor只支持2-D和3-D的矩阵，但是SparseCooTensor支持多维的Tensor，非零元素可以是Tensor。这里考虑到确实是csr比较好实现，可以考虑自定以一个转换方法，把这种非零元素是Tensor的coo转换成csr，然后再进行计算。

zkh2016 reviewed (commented):
另外建议可以再写下GPU实现。

gsq7474741 commented:
> 另外建议可以再写下GPU实现。

好的，GPU的我黑客松之后再提个pr

gsq7474741 committed:
add sparse elementwise v0.8

gsq7474741 committed:
add sparse elementwise v0.8

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,432 @@
+namespace phi {
+namespace sparse {
+
+#define DEFINE_CSR_ELEMENTWISE_KERNEL(name)                                    \
```
这里的作用看上去就是替换函数名字合functor的名字，可以试着把计算逻辑单独放一个函数，这个宏里面调用该函数，也可以把functor单参数传递进来。

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,432 @@
+    const auto& n_row = x.dims().size() == 2 ? x.dims()[0] : x.dims()[1];      \
+    const auto& n_col = x.dims().size() == 2 ? x.dims()[1] : x.dims()[2];      \
+    const auto& x_nnz = x.non_zero_elements().numel();                         \
+    const auto* x_crows_data = x.non_zero_crows().data<int64_t>();             \
```
这里的crows，cols需要支持模板，kernel内可以用二次分发来获取坐标/索引的dtype，可以参考其他计算类kernel的实现。

gsq7474741 replied:
```
  const auto& csr_crows = x.non_zero_crows();
  const auto& csr_cols = x.non_zero_cols();
  const auto& csr_values = x.non_zero_elements();
  const int64_t* csr_crows_data = csr_crows.data<int64_t>();
  const int64_t* csr_cols_data = csr_cols.data<int64_t>();
```
是指这个写法嘛

zkh2016 replied:
参考这里哈：
[  PD_VISIT_INTEGRAL_TYPES二次分发](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/sparse/cpu/sparse_mask_kernel.cc#L82)

[获取索引信息]https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/sparse/cpu/sparse_mask_kernel.cc#L56

gsq7474741 replied:
好滴 我看看

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,432 @@
+    out_batch_crows_vec.push_back(0);                                          \
+    int64_t x_prev_batch_nnz = 0;                                              \
+    int64_t y_prev_batch_nnz = 0;                                              \
+    for (int b = 0; b < n_batch; b++) {                                        \
```
这里的计算逻辑可以加点注释。

zkh2016 replied:
这里的方式是把输入当作无序的？我们当前的SparseCooTensor构造的时候会进行排序和合并重复项，参考[PR#41784](https://github.com/PaddlePaddle/Paddle/pull/41784)，所以这里建议思考下考换种算法实现。

gsq7474741 replied:
已更新

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,432 @@
+  out_batch_crows_vec.push_back(0);
+  int64_t x_prev_batch_nnz = 0;
+  int64_t y_prev_batch_nnz = 0;
+  for (int b = 0; b < n_batch; b++) {
```
这部分看着不少相似代码，可以试着提取出来作为一个函数调用。

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	python/paddle/sparse/__init__.py
#	python/paddle/utils/code_gen/sparse_bw_api.yaml

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
add sparse elementwise v1.0

gsq7474741 committed:
add sparse elementwise v1.0

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,550 @@
+        if (A_j == B_j) {
+          T result = functor(x_values_data[x_idx + x_prev_batch_nnz],
+                             y_values_data[y_idx + y_prev_batch_nnz]);
+          if (result != 0) {
```
这个代码块重复的太多了，可以提取出来。

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,550 @@
+  IntT a = 0;
+  IntT b = 0;
+  //  merge x and y
+  for (int i = 0; i < std::max(nnz_x, nnz_y); ++i) {
```
这里merge的整体逻辑和csr merge一行的逻辑是一样的，可以考虑下把merge的逻辑写成通用的函数。

gsq7474741 replied:
已更新

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,550 @@
+    }
+  }
+
+  if (nnz == 0) {
```
输出为空的情况？可以试试*out = SparseCoTensor()

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,550 @@
+DEFINE_CSR_ELEMENTWISE_KERNEL(Add)
+DEFINE_CSR_ELEMENTWISE_KERNEL(Subtract)
+DEFINE_CSR_ELEMENTWISE_KERNEL(Multiply)
+DEFINE_CSR_ELEMENTWISE_KERNEL(Divide)
```
除法需要把零值考虑进去，这个kernrl貌似并不太适用，可以把merge逻辑扩展下，适配下除法的情况。

gsq7474741 committed:
add sparse elementwise v1.1

gsq7474741 committed:
add sparse elementwise v1.1

gsq7474741 committed:
Merge branch 'PaddlePaddle:develop' into add_sparse_add

gsq7474741 committed:
add sparse elementwise v1.1

gsq7474741 committed:
add sparse elementwise v1.2

gsq7474741 committed:
add sparse elementwise v1.2

zkh2016 commented on paddle/phi/kernels/funcs/elementwise_functor.h:
```diff
@@ -50,6 +76,19 @@ template <typename T>
+template <typename T>
+struct MultiplyPtrFunctor {
+  inline HOSTDEVICE std::vector<T> operator()(const T* a,
+                                              const T* b,
```
这些functor计算模式都一样，可以考虑只实现一个函数，然后把计算的functor传进来，函数内循环调用functor(a[i], b[i])。
看这里返回vector的目的是后面判断是否为0，可以考虑只返回一个bool is_zero，循环里把functor的计算结果是否为0于is_zero进行或运算，也就是只要有一个不为0就返回的is_zero=false，否则返回的is_zero=true。

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,409 @@
+                              const SparseCooTensor& y,
+                              SparseCooTensor* out,
+                              const Functor& functor) {
+  PADDLE_ENFORCE_EQ(x.dims(),
```
注释和错误信息里说明下，当前只支持相同shape的计算，不支持broadcast的计算

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,409 @@
+                        x.dims(),
+                        y.dims()));
+  int64_t element_size = 1;
+  for (auto j = 1; j < x.non_zero_elements().dims().size(); ++j) {
```
如果x的shape是[2,3,4,5], 前两维是坐标呢？

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,409 @@
+
+  std::vector<IntT> out_indexs;
+  std::vector<T> out_values_vec;
+  out_indexs.reserve(max_len);
```
这里的max_len最大等于x.nnz() + y.nnz()吧？

gsq7474741 replied:
这里已修改为在除法时才为max_len

zkh2016 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,409 @@
+          functor(a_values + a * el_len, b_values[b_index[b]], el_len);
+      if (!is_zero(result.data(), result.size())) {
+        c_index[nnz] = a_index[a];
+        memcpy(
```
考虑到c_values的空间已经按最大需要分配好，建议直接把c传到计算函数进去直接把计算结果写入到c_values，这里只根据is_zero控制是否++nnz就可以了

gsq7474741 replied:
这里全为0则不写入c

gsq7474741 committed:
add sparse elementwise v1.3

gsq7474741 committed:
add sparse elementwise v1.3

gsq7474741 commented:
**暂不拆解**
```
本PR主体功能已完善，拆解为
- PR 1：#42374 
- PR 2：#42375

进行merge
```

zkh2016 commented on python/paddle/fluid/tests/unittests/test_sparse_elementwise_op.py:
```diff
@@ -0,0 +1,100 @@
+                dense_x = paddle.to_tensor(x).astype(dtype)
+                dense_y = paddle.to_tensor(y).astype(dtype)
+                coo_x = dense_x.to_sparse_coo(2)
+                coo_y = dense_y.to_sparse_coo(2)
```
非0元素是标量和tensor的都分别测下，另外coo需要测试反向。

gsq7474741 replied:
好的

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

gsq7474741 committed:
add sparse elementwise v1.5

gsq7474741 committed:
add sparse elementwise v2.0

add sparse elementwise v2.0

add sparse elementwise v2.0

test=document_fix

fix math.py

fix code gen yaml

fix math.py

fix math.py

TCChenlong commented:
请在 Describe 中添加 RFC & 中文文档的链接

TCChenlong commented on python/paddle/sparse/math.py:
```diff
@@ -0,0 +1,207 @@
+
+def add(x, y, name=None):
+    """
+    Add two sparse tensors element-wise. The equation is:
```
这里的描述和中文保持一致吧，那里的信息比较全
其他几个也是

gsq7474741 replied:
done

gsq7474741 commented:
> 请在 Describe 中添加 RFC & 中文文档的链接

已贴链接

zkh2016 commented on python/paddle/fluid/tests/unittests/test_sparse_elementwise_op.py:
```diff
@@ -0,0 +1,148 @@
+
+from __future__ import print_function
+import unittest
+from operator import __add__, __sub__, __mul__, __truediv__
```
div的命名为什么和其他的不一样？

gsq7474741 replied:
导入这个只是为了能够在计算dense的参照结果时复用op(x, y)这句话而已

![6433a7e91033e5fa1c8ea84bb69cf7c8](https://user-images.githubusercontent.com/38883252/166092634-8e5ff076-c6c1-4233-a108-d42a4e5d9e82.png)

TCChenlong reviewed (dismissed):
LGTM

DDDivano commented on python/paddle/fluid/tests/unittests/test_sparse_elementwise_op.py:
```diff
@@ -0,0 +1,148 @@
+                        equal_nan=True))
+
+    def test_coo_add(self):
+        if paddle.device.get_device() == "cpu":
```
如果只实现了CPU的版本，需要补充一个异常的Case，在GPU下调用该方法需要有一个友好的异常提示说明当前API不支持GPU硬件调用。

gsq7474741 replied:
已更新

gsq7474741 replied:
加在函数本身里面了，后期如果增加GPU版本只需要在测试文件和math.py里删除两个判断即可

gsq7474741 replied:
@DDDivano

gsq7474741 commented:
已更新 @DDDivano

zkh2016 commented on python/paddle/sparse/math.py:
```diff
@@ -0,0 +1,219 @@
+        out = x + y
+
+    Args:
+        x (Tensor): the input tensor, it's data type should be float32, float64.
```
这里只支持float32和float64？`paddle.add`是支持float32、float64、int32、int64，我看kernel注册的地方是都支持的，这边补充下，另外单测也可以补充下。

gsq7474741 replied:
ok 我改下

gsq7474741 replied:
done

gsq7474741 committed:
working for fix int div

gsq7474741 committed:
fix int div

gsq7474741 committed:
fix int div

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	python/paddle/sparse/__init__.py
#	python/paddle/utils/code_gen/sparse_api.yaml
#	python/paddle/utils/code_gen/sparse_bw_api.yaml

gsq7474741 committed:
fix int div

gsq7474741 committed:
fix int div docstring

gsq7474741 committed:
fix coverage

gsq7474741 committed:
move to incubate

gsq7474741 committed:
move to incubate

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	paddle/phi/kernels/cpu/activation_kernel.cc
#	python/paddle/utils/code_gen/sparse_bw_api.yaml

gsq7474741 committed:
move to incubate

gsq7474741 committed:
move to incubate

gsq7474741 committed:
move to incubate

gsq7474741 committed:
fix incubate import

Null

gsq7474741 committed:
move to incubate

gsq7474741 committed:
move to incubate

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	python/paddle/incubate/__init__.py
#	python/setup.py.in

TCChenlong reviewed (dismissed):
LGTM

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	paddle/phi/tests/kernels/CMakeLists.txt
#	python/paddle/incubate/sparse/__init__.py
#	python/paddle/incubate/sparse/nn/__init__.py

gsq7474741 committed:
move to incubate

gsq7474741 committed:
resolve conflict

gsq7474741 committed:
resolve conflict

gsq7474741 committed:
pre-commit fix

zkh2016 commented on python/setup.py.in:
```diff
@@ -281,6 +281,7 @@ packages=['paddle',
           'paddle.incubate.tensor',
           'paddle.incubate.multiprocessing',
           'paddle.incubate.nn',
+          'paddle.incubate.sparse',
```
这个已经加过了，后面再提个PR去掉下

gsq7474741 replied:
ok

TCChenlong reviewed (dismissed):
LGTM

XiaoguangHu01 commented on paddle/phi/kernels/cpu/activation_kernel.cc:
```diff
@@ -182,6 +183,15 @@ PD_REGISTER_ACTIVATION_KERNEL(swish, SwishKernel)
 PD_REGISTER_ACTIVATION_KERNEL(round, RoundKernel)
 PD_REGISTER_ACTIVATION_KERNEL(floor, FloorKernel)
 PD_REGISTER_ACTIVATION_KERNEL(ceil, CeilKernel)
+PD_REGISTER_KERNEL(opposite,
```
这里是取负号操作吗？这个名字是否是业界普遍的叫法？是否应该叫negative呢

gsq7474741 replied:
已修改

XiaoguangHu01 commented on python/paddle/incubate/sparse/math.py:
```diff
@@ -0,0 +1,248 @@
+    ..  code-block:: python
+
+        import paddle
+        from paddle.fluid.framework import _test_eager_guard
```
这里为什么需要调用内部的api呢？默认是imperative模式，这里只在eager模式下才能生效吗？

zkh2016 replied:
嗯，当前Sparse只在eager模式下生效。

XiaoguangHu01 commented on python/paddle/utils/code_gen/sparse_bw_api.yaml:
```diff
@@ -53,6 +95,20 @@
+  args : (Tensor x, Tensor y, Tensor out_grad)
+  output : Tensor(x_grad), Tensor(y_grad)
+  kernel :
+    func : sparse_elementwise_div_grad_csr{sparse_csr, sparse_csr, sparse_csr -> sparse_csr, sparse_csr}
```
div_grad -> sub_grad
为什么单测没有发现这个错误呢？请补充一下单测

gsq7474741 replied:
刚开始csr不支持反向，所以没测，现已修改

XiaoguangHu01 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_grad_kernel.cc:
```diff
@@ -0,0 +1,413 @@
+}  // namespace sparse
+}  // namespace phi
+
+PD_REGISTER_KERNEL(sparse_elementwise_add_grad_csr,
```
kenerl的名字命名方式，对齐其他算子sparse_csr_add_grad

XiaoguangHu01 commented on paddle/phi/kernels/sparse/cpu/sparse_elementwise_kernel.cc:
```diff
@@ -0,0 +1,451 @@
+  kernel->InputAt(1).SetDataLayout(phi::DataLayout::SPARSE_CSR);
+}
+
+PD_REGISTER_KERNEL(sparse_elementwise_sub_coo,
```
sub 用全称，跟python api的命名一致

gsq7474741 replied:
已修改，均已对齐，sub div mul -> subtract divide multiply，感谢指正

zhwesky2010 commented:
你好，yaml形式规范可以参考下这个PR，合并到一起在设计上更好：https://github.com/PaddlePaddle/Paddle/pull/42935/files#diff-bfad6f19739eb95b60bd46fc6a77adbf3ccb4a5ebac465b7aba845ae15470f11

![infoflow 2022-06-14 16-42-07](https://user-images.githubusercontent.com/52485244/173534197-ba32ffca-6928-4fb5-9480-0b428bb1d253.png)

gsq7474741 committed:
Merge remote-tracking branch 'origin/develop' into add_sparse_add

# Conflicts:
#	python/paddle/incubate/sparse/__init__.py
#	python/paddle/utils/code_gen/sparse_bw_api.yaml

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 committed:
kernel and api name change

gsq7474741 commented:
> 你好，yaml形式规范可以参考下这个PR，合并到一起在设计上更好：https://github.com/PaddlePaddle/Paddle/pull/42935/files#diff-bfad6f19739eb95b60bd46fc6a77adbf3ccb4a5ebac465b7aba845ae15470f11
> 
> ![infoflow 2022-06-14 16-42-07](https://user-images.githubusercontent.com/52485244/173534197-ba32ffca-6928-4fb5-9480-0b428bb1d253.png)

好的，已参考该设计修改

zhwesky2010 commented:
你好，经过内部讨论，目前形成以下一些sparse开发规范要求：

1.【yaml规则】：写到同一个yaml api，不要写多个
2.【kernel名规则】：[计算名] + 异构后缀，matmul、matmul_csr_dense、softmax、softmax_csr
3.【文件名规则】：sparse/xx_kernel.cc，sparse/xx_kernel.cu，以目录区分，文件名与dense一致

辛苦对kenrel注册名、与文件名进行修改

gsq7474741 commented:
> 你好，经过内部讨论，目前达成以下一些sparse开发规范要求：
> 
> 1.【yaml规则】：写到同一个yaml api，不要写多个，yaml需支持调度 2.【kernel名规则】：[计算名] + 异构后缀，matmul、matmul_csr_dense、softmax、softmax_csr 3.【文件名规则】：sparse/xx_kernel.cc，sparse/xx_kernel.cu，以目录区分，文件名与dense一致
> 
> 辛苦对kenrel注册名、与文件名进行修改

加减乘除要写成四个文件么

zhwesky2010 commented:
> > 你好，经过内部讨论，目前达成以下一些sparse开发规范要求：
> > 1.【yaml规则】：写到同一个yaml api，不要写多个，yaml需支持调度 2.【kernel名规则】：[计算名] + 异构后缀，matmul、matmul_csr_dense、softmax、softmax_csr 3.【文件名规则】：sparse/xx_kernel.cc，sparse/xx_kernel.cu，以目录区分，文件名与dense一致
> > 辛苦对kenrel注册名、与文件名进行修改
> 
> 加减乘除要写成四个文件么

就写到sparse/elementwise_kernel.h/cc/cu吧，文件不用sparse前缀，使用目录区分

gsq7474741 commented:
> > > 你好，经过内部讨论，目前达成以下一些sparse开发规范要求：
> 
> > > 1.【yaml规则】：写到同一个yaml api，不要写多个，yaml需支持调度 2.【kernel名规则】：[计算名] + 异构后缀，matmul、matmul_csr_dense、softmax、softmax_csr 3.【文件名规则】：sparse/xx_kernel.cc，sparse/xx_kernel.cu，以目录区分，文件名与dense一致
> 
> > > 辛苦对kenrel注册名、与文件名进行修改
> 
> > 
> 
> > 加减乘除要写成四个文件么
> 
> 
> 
> 这里复用比较多，就合到elementwise_kernel.h/cc/cu吧，不用sparse前缀

指的是dense的这个文件么

zhwesky2010 commented:
就写到sparse/elementwise_kernel.h/cc/cu里吧，文件不用sparse前缀，使用目录区分

gsq7474741 committed:
kernel and api name change

zhwesky2010 commented on python/paddle/utils/code_gen/sparse_api.yaml:
```diff
@@ -1,3 +1,12 @@
+  args : (Tensor x, Tensor y)
+  output : Tensor(out)
+  kernel :
+    func : add_coo{sparse_coo -> sparse_coo},
```
这里二元的kernel名是 add_coo_coo，因为后面我们还会补充add_coo_dense的情形

gsq7474741 replied:
好的

zhwesky2010 commented on python/paddle/utils/code_gen/sparse_bw_api.yaml:
```diff
@@ -1,3 +1,11 @@
+  args : (Tensor x, Tensor y, Tensor out_grad)
+  output : Tensor(x_grad), Tensor(y_grad)
+  kernel :
+    func : add_grad_coo{sparse_coo, sparse_coo, sparse_coo -> sparse_coo, sparse_coo},
```
反向kernel名 add_coo_coo_grad

gsq7474741 committed:
fix kernel name

zkh2016 reviewed (approved):
LGTM

TCChenlong reviewed (approved):
LGTM

XiaoguangHu01 reviewed (approved):
LGTM

zhwesky2010 commented:
@gsq7474741 你好，这个kernel在`FLAGS_use_system_allocator=0`时单测是正确的，在`FLAGS_use_system_allocator=1`时会发生错误，kernel可能有地址访问bug需要修复。

![586980951007279ae290208c1332e0f2](https://user-images.githubusercontent.com/52485244/197480802-538e11db-2c79-4fe9-8639-896f466de58a.png)

![2fdac3d7babb941612da7827f697cdf3](https://user-images.githubusercontent.com/52485244/197480789-67b10c21-449c-4985-ae99-e29d70ee597b.png)
